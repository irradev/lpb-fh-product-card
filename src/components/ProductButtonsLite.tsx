
import React, { CSSProperties } from 'react';
import styles from  '../styles/styles.module.css';

export interface Props {
  className?: string;
  style?: CSSProperties;
  counter: number;
  isMaxCountReached: boolean;
  increaseBy: () => void;
  decreaseBy: () => void;
  reset: () => void;
  
}

export const ProductButtonsLite = ({counter, isMaxCountReached, increaseBy, decreaseBy, reset, style={}, className= ''}:Props) => {
  return (
    <div 
      className={`${styles.buttonsContainer} ${className}`}
      style={style}
    >
      <button 
        className={styles.buttonMinus}
        onClick={decreaseBy}
      >
        -
      </button>

      <div className={styles.countLabel}>{counter}</div>

      <button
        disabled={isMaxCountReached}
        className={`${styles.buttonAdd} ${isMaxCountReached && styles.disabled}`}
        onClick={increaseBy}
        >
        +
      </button>

      <button
        className={`${styles.buttonAdd}`}
        style={{fontSize: '14px', width: '70px', borderRadius:'8px', marginLeft:'15px'}}
        onClick={reset}
        >
        Reset
      </button>
    </div>
  )
}
