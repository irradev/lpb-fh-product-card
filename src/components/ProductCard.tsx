import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';

import { ProductImage, ProductTitle, ProductButtons} from "../components/"

import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

import styles from  '../styles/styles.module.css';


export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;



export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element; 
  className?: string;
  style?: CSSProperties;
  value?: number;
  initialValues?: InitialValues;
  onChange?: (args: onChangeArgs) => void;
}

export const ProductCard = ({ 
  children, 
  product, 
  className = '', 
  style, 
  value, 
  initialValues,  
  onChange 
}: Props) => {
  const { counter, maxCount, isMaxCountReached, increaseBy, reset } = useProduct( { product, value, initialValues, onChange } );

  const data: ProductContextProps = {
    counter,
    maxCount,
    product,
    increaseBy,
  }

  return (
    <Provider value={data}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached: isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset,

        })}
      </div>
    </Provider>
  )

  // return (
  //   <Provider value={data}>
  //     <div className={styles.productCard}>
  //       <ProductImage img={product.img} />
  //       <ProductTitle title={product.title}/>
  //       <ProductButtons counter={counter} increaseBy={increaseBy} />
  //     </div>
  //   </Provider>
  // )
}

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;