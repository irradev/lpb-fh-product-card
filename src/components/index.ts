import { ProductCardHOCProps } from "../interfaces/interfaces";

import { ProductCard as ProductCardHOC } from "./ProductCard";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductButtons } from "./ProductButtons";
import { ProductButtonsLite } from "./ProductButtonsLite";

export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";
export { ProductButtons } from "./ProductButtons";
export { ProductButtonsLite } from "./ProductButtonsLite";

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
  ButtonsLite: ProductButtonsLite,
});

export default ProductCard;
