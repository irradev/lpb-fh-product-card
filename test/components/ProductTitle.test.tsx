import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components';

import { product1 } from '../data/products';

describe('ProductTitle', () => {

	test('debe mostrar correctamente el titulo personalizado', () => {
		const wrapper = renderer.create(
			<ProductTitle title="Custom Product" />
		)

		expect(wrapper.toJSON()).toMatchSnapshot();
	});

	test('debe mostrar el nombre del producto por defecto', () => {
		const wrapper = renderer.create(
			<ProductCard product={product1}>
				{
					() => (
						<ProductTitle />
					)
				}
			</ProductCard>
		)

		expect(wrapper.toJSON()).toMatchSnapshot();
	})

})