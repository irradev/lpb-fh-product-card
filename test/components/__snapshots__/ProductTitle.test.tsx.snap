// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostrar correctamente el titulo personalizado 1`] = `
<span
  className="productDescription "
>
  Custom Product
</span>
`;

exports[`ProductTitle debe mostrar el nombre del producto por defecto 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Taza Developer - Sin Imagen
  </span>
</div>
`;
